New template merge_response

gjohansson-ST opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
# Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Implements a new template `merge_response` which is to consolidate responses from service calls.
See discussion in https://github.com/home-assistant/architecture/discussions/1045 for context.

```python
def merge_response(value: ServiceResponse) -> list[Any]:
    """Merge action responses into single list.

    Checks that the input is a correct service response:
    {
        "entity_id": {str: dict[str, Any]},
    }
    If response is a single list, it will extend the list with the items
        and add the entity_id and value_key to each dictionary for reference.
    If response is a dictionary or multiple lists,
        it will append the dictionary/lists to the list
        and add the entity_id to each dictionary for reference.
    """
```

#### A single dict item per entity response

Action call response:

```json
{
    "calendar.sports": {
        "events": [
            {
                "start": "2024-02-27T17:00:00-06:00",
                "end": "2024-02-27T18:00:00-06:00",
                "summary": "Basketball vs. Rockets",
                "description": "",
            }
        ]
    },
    "calendar.local_furry_events": {"events": []},
    "calendar.yap_house_schedules": {
        "events": [
            {
                "start": "2024-02-26T08:00:00-06:00",
                "end": "2024-02-26T09:00:00-06:00",
                "summary": "Dr. Appt",
                "description": "",
            },
            {
                "start": "2024-02-28T20:00:00-06:00",
                "end": "2024-02-28T21:00:00-06:00",
                "summary": "Bake a cake",
                "description": "something good",
            },
        ]
    },
}
```

Template: `{{ merge_response(response_variable) }}`

Result: 

```json
[
  {
    "description": "",
    "end": "2024-02-27T18:00:00-06:00",
    "entity_id": "calendar.sports",
    "start": "2024-02-27T17:00:00-06:00",
    "summary": "Basketball vs. Rockets",
    "value_key": "events"
  },
  {
    "description": "",
    "end": "2024-02-26T09:00:00-06:00",
    "entity_id": "calendar.yap_house_schedules",
    "start": "2024-02-26T08:00:00-06:00",
    "summary": "Dr. Appt",
    "value_key": "events"
  },
  {
    "description": "something good",
    "end": "2024-02-28T21:00:00-06:00",
    "entity_id": "calendar.yap_house_schedules",
    "start": "2024-02-28T20:00:00-06:00",
    "summary": "Bake a cake",
    "value_key": "events"
  }
]
```

#### A single non dict item per entity response

Action call response:

```json
  {
    "binary_sensor.workday": {
      "workday": true,
    },
    "binary_sensor.workday2": {
      "workday": false,
    },
  }
```

Template: `{{ merge_response(response_variable) }}`

Result: 
```json
  [
    {
      "entity_id": "binary_sensor.workday",
      "workday": true,
    },
    {
      "entity_id": "binary_sensor.workday2",
      "workday": false,
    },
  ]
```

#### Multiple items per entity response

Action call response:

```json
  {
    "vacuum.deebot_n8_plus_1": {
      "header": {
        "ver": "0.0.1",
      },
      "payloadType": "j",
      "resp": {
        "body": {
          "msg": "ok",
        },
      },
    },
    "vacuum.deebot_n8_plus_2": {
      "header": {
        "ver": "0.0.1",
      },
      "payloadType": "j",
      "resp": {
        "body": {
          "msg": "ok",
        },
      },
    },
  }
```

Template: `{{ merge_response(response_variable) }}`

Result: 
```json
  [
    {
      "entity_id": "vacuum.deebot_n8_plus_1",
      "header": {
        "ver": "0.0.1",
      },
      "payloadType": "j",
      "resp": {
        "body": {
          "msg": "ok",
        },
      },
    },
    {
      "entity_id": "vacuum.deebot_n8_plus_2",
      "header": {
        "ver": "0.0.1",
      },
      "payloadType": "j",
      "resp": {
        "body": {
          "msg": "ok",
        },
      },
    },
  ]
```

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/34050

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

gjohansson-ST committed:
New template merge_response

emontnemery commented:
A few comments:
- The proposed function works for the output from `calendar.get_events`, but what about other services?
- Since the output is a list, it should probably be possible for the user to specify how to sort the output
- We may need an option to specify the expected key of the lists to append

gjohansson-ST committed:
Extending

gjohansson-ST commented:
> * The proposed function works for the output from `calendar.get_events`, but what about other services?

Added some more examples in the tests (but there is limited examples)
This should now deal with more than just lists (any type valid for service response should be fine)

> * Since the output is a list, it should probably be possible for the user to specify how to sort the output

Added sorting, will expand it to also deal with something else than dict keys

> * We may need an option to specify the expected key of the lists to append

I understood this as extracting only a single key from a resulting dict so it has been added as `single_key`

> * We may want to add the source entity_id to each item to make sense of the output

I don't see the use of that?

I wonder a bit about the use of sorting and single_key as both of these are available themselves in templates so are we not just simply making this far more complex than what's normal for templates (keep it simple)?

gjohansson-ST committed:
Extend comment

emontnemery commented on homeassistant/helpers/template.py:
```diff
@@ -2113,6 +2115,42 @@ def as_timedelta(value: str) -> timedelta | None:
+) -> list[Any]:
+    """Merge service responses into single list.
+
+    Checks that the input is a correct service response:
+    {
+        "entity_id": {str: Any},
+    }
```
I think the docstring should explain the special treatment of lists too

gjohansson-ST replied:
I added a line for explaining the list treatment.

emontnemery commented on homeassistant/helpers/template.py:
```diff
@@ -2113,6 +2115,42 @@ def as_timedelta(value: str) -> timedelta | None:
+    {
+        "entity_id": {str: Any},
+    }
+    Returns empty list by default.
```
The function returns the empty list *in case of any error*, is that wanted instead of erroring out? Why don't we want to blow up instead?

gjohansson-ST replied:
I changed this so it blows.

gjohansson-ST committed:
Update

gjohansson-ST committed:
Fixes

gjohansson-ST committed:
Fix comments

emontnemery commented on homeassistant/helpers/template.py:
```diff
@@ -2112,6 +2113,65 @@ def as_timedelta(value: str) -> timedelta | None:
+                        dict_in_list["value_key"] = value_key
+                response_items.extend(type_response)
+            else:
+                break  # Don't continue looping if not a list
```
Please explain why we break out of the `for value_key, type_response in entity_response.items()` loop

gjohansson-ST replied:
Provided explanation (as logic is handled below for when it's not a single list with dicts.

gjohansson-ST committed:
Mods

gjohansson-ST committed:
snapshots

gjohansson-ST commented on homeassistant/helpers/template.py:
```diff
@@ -2112,6 +2113,65 @@ def as_timedelta(value: str) -> timedelta | None:
+            response_items.append(_response)
+
+    # Allows to sort the response items by a key within the list
+    if sort_by and isinstance(response_items[0], dict):
```
This is already guarded that it needs to be a dict and also below it will raise if the key provided does not exist in the dict.

gjohansson-ST commented on tests/helpers/test_template.py:
```diff
@@ -6236,3 +6237,416 @@ async def test_template_thread_safety_checks(hass: HomeAssistant) -> None:
+async def test_merge_response(
+    hass: HomeAssistant,
+    service_response: dict,
+    snapshot: SnapshotAssertion,
```
Both response and render is now in the snapshot file (sorted)

gjohansson-ST commented on homeassistant/helpers/template.py:
```diff
@@ -2112,6 +2113,65 @@ def as_timedelta(value: str) -> timedelta | None:
+        if not isinstance(entity_response, dict):
+            raise TypeError("Response is not a dictionary")
+        for value_key, type_response in entity_response.items():
+            if len(entity_response) == 1 and isinstance(type_response, list):
```
Comment provided but I'm not entirely sure it's self-explanatory enough for an outsider to understand 🤷‍♂️

gjohansson-ST commented:
Docs updated to reflect the changes https://deploy-preview-34050--home-assistant-docs.netlify.app/docs/configuration/templating/#merge-action-responses

gjohansson-ST committed:
Fixes from discussion

gjohansson-ST commented:
As discussed the use of sorting and using selected key has been removed

emontnemery reviewed (approved):
Thanks, @gjohansson-ST 👍

gjohansson-ST commented:
Test failures are not related to this PR. Merging
